File: src/main/java/io/outfoxx/typescriptpoet/ClassSpec.kt
/*
 * Copyright 2017 Outfox, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.outfoxx.typescriptpoet


/** A generated `class` declaration. */
class ClassSpec
private constructor(
   builder: Builder
) {

  val name = builder.name
  val javaDoc = builder.javaDoc.build()
  val decorators = builder.decorators.toImmutableList()
  val modifiers = builder.modifiers.toImmutableSet()
  val typeVariables = builder.typeVariables.toImmutableList()
  val superClass = builder.superClass
  val mixins = builder.mixins.toImmutableList()
  val propertySpecs = builder.propertySpecs.toImmutableList()
  val constructor = builder.constructor
  val functionSpecs = builder.functionSpecs.toImmutableList()

  internal fun emit(codeWriter: CodeWriter) {

    val constructorProperties: Map<String, PropertySpec> = constructorProperties()

    codeWriter.emitJavaDoc(javaDoc)
    codeWriter.emitDecorators(decorators, false)
    codeWriter.emitModifiers(modifiers, setOf(Modifier.PUBLIC))
    codeWriter.emit("class")
    codeWriter.emitCode(" %L", name)
    codeWriter.emitTypeVariables(typeVariables)

    val superClass = if (superClass != null) CodeBlock.of("extends %T",
                                                                                              superClass)
    else CodeBlock.empty()
    val mixins = mixins.map { CodeBlock.of("%T", it) }.let {
      if (it.isNotEmpty()) it.joinToCode(prefix = "implements ") else CodeBlock.empty()
    }

    if (superClass.isNotEmpty() && mixins.isNotEmpty()) {
      codeWriter.emitCode(" %L %L", superClass, mixins)
    }
    else if (superClass.isNotEmpty() || mixins.isNotEmpty()) {
      codeWriter.emitCode(" %L%L", superClass, mixins)
    }

    codeWriter.emit(" {\n")
    codeWriter.indent()

    // Non-static properties.
    for (propertySpec in propertySpecs) {
      if (constructorProperties.containsKey(propertySpec.name)) {
        continue
      }
      codeWriter.emit("\n")
      propertySpec.emit(codeWriter, setOf(Modifier.PUBLIC), asStatement = true)
    }

    // Write the constructor manually, allowing the replacement
    // of property specs with constructor parameters
    constructor?.let {
      codeWriter.emit("\n")

      if (it.decorators.isNotEmpty()) {
        codeWriter.emit(" ")
        codeWriter.emitDecorators(it.decorators, true)
        codeWriter.emit("\n")
      }

      if (it.modifiers.isNotEmpty()) {
        codeWriter.emitModifiers(it.modifiers)
      }

      codeWriter.emit("constructor")

      var body = constructor.body

      // Emit constructor parameters & property specs that can be replaced with parameters
      it.parameters.emit(codeWriter, rest = it.restParameter) { param, isRest ->
        var property = constructorProperties[param.name]
        if (property != null && !isRest) {

          // Ensure the parameter always has a modifier (that makes it a property in TS)
          if (property.modifiers.none {
            it.isOneOf(Modifier.PUBLIC, Modifier.PRIVATE,
                       Modifier.PROTECTED, Modifier.READONLY)
          }) {
            // Add default public modifier
            property = property.toBuilder().addModifiers(Modifier.PUBLIC).build()
          }
          property.emit(codeWriter, setOf(), withInitializer = false)
          param.emitDefaultValue(codeWriter)

          // Remove initializing statements

          val bodyBuilder = body.toBuilder()
          bodyBuilder.remove(constructorPropertyInitSearch(property.name))
          body = bodyBuilder.build()
        }
        else {
          param.emit(codeWriter, isRest = isRest)
        }
      }

      codeWriter.emit(" {\n")
      codeWriter.indent()
      codeWriter.emitCode(body)
      codeWriter.unindent()
      codeWriter.emit("}\n")
    }

    // Constructors.
    for (funSpec in functionSpecs) {
      if (!funSpec.isConstructor) continue
      codeWriter.emit("\n")
      funSpec.emit(codeWriter, name, setOf(Modifier.PUBLIC))
    }

    // Functions (static and non-static).
    for (funSpec in functionSpecs) {
      if (funSpec.isConstructor) continue
      codeWriter.emit("\n")
      funSpec.emit(codeWriter, name, setOf(Modifier.PUBLIC))
    }

    codeWriter.unindent()

    if (!hasNoBody) {
      codeWriter.emit("\n")
    }
    codeWriter.emit("}\n")
  }

  /** Returns the properties that can be declared inline as constructor parameters. */
  private fun constructorProperties(): Map<String, PropertySpec> {
    val result: MutableMap<String, PropertySpec> = LinkedHashMap()
    val body = constructor?.body?.toString() ?: return result
    for (property in propertySpecs) {
      val parameter = constructor.parameter(property.name) ?: continue
      if (parameter.type != property.type) continue
      if (parameter.optional != property.optional) continue
      if (property.initializer != null) continue
      if (!body.contains(constructorPropertyInitSearch(property.name))) continue
      result[property.name] = property
    }
    return result
  }

  private fun constructorPropertyInitSearch(name: String): Regex {
    return """(\A|\n|;)\s*\Qthis.$name = $name\E[ \t\x0B\f\r]*(\n|;|\z)""".toRegex()
  }

  private val hasNoBody: Boolean
    get() {
      if (propertySpecs.isNotEmpty()) {
        val constructorProperties = constructorProperties()
        propertySpecs
           .filterNot { constructorProperties.containsKey(it.name) }
           .forEach { return false }
      }
      return constructor == null && functionSpecs.isEmpty()
    }

  fun toBuilder(): Builder {
    val builder = Builder(name)
    builder.javaDoc.add(javaDoc)
    builder.decorators += decorators
    builder.modifiers += modifiers
    builder.typeVariables += typeVariables
    builder.superClass = superClass
    builder.mixins += mixins
    builder.propertySpecs += propertySpecs
    builder.constructor = constructor
    builder.functionSpecs += functionSpecs
    return builder
  }

  class Builder(
     internal val name: String
  ) {

    internal val javaDoc = CodeBlock.builder()
    internal val decorators = mutableListOf<DecoratorSpec>()
    internal val modifiers = mutableListOf<Modifier>()
    internal val typeVariables = mutableListOf<TypeName.TypeVariable>()
    internal var superClass: TypeName? = null
    internal val mixins = mutableListOf<TypeName>()
    internal val propertySpecs = mutableListOf<PropertySpec>()
    internal var constructor: FunctionSpec? = null
    internal val functionSpecs = mutableListOf<FunctionSpec>()

    fun addJavadoc(format: String, vararg args: Any) = apply {
      javaDoc.add(format, *args)
    }

    fun addJavadoc(block: CodeBlock) = apply {
      javaDoc.add(block)
    }

    fun addDecorators(decoratorSpecs: Iterable<DecoratorSpec>) = apply {
      decorators += decoratorSpecs
    }

    fun addDecorator(decoratorSpec: DecoratorSpec) = apply {
      decorators += decoratorSpec
    }

    fun addModifiers(vararg modifiers: Modifier) = apply {
      this.modifiers += modifiers
    }

    fun addTypeVariables(typeVariables: Iterable<TypeName.TypeVariable>) = apply {
      this.typeVariables += typeVariables
    }

    fun addTypeVariable(typeVariable: TypeName.TypeVariable) = apply {
      typeVariables += typeVariable
    }


    fun superClass(superClass: TypeName) = apply {
      check(this.superClass == null) { "superclass already set to ${this.superClass}" }
      this.superClass = superClass
    }

    fun addMixins(mixins: Iterable<TypeName>) = apply {
      this.mixins += mixins
    }

    fun addMixin(mixin: TypeName) = apply {
      mixins += mixin
    }

    fun constructor(constructor: FunctionSpec?) = apply {
      if (constructor != null) {
        require(constructor.isConstructor) {
          "expected a constructor but was ${constructor.name}; use FunctionSpec.constructorBuilder when building"
        }
      }
      this.constructor = constructor
    }

    fun addProperties(propertySpecs: Iterable<PropertySpec>) = apply {
      this.propertySpecs += propertySpecs
    }

    fun addProperty(propertySpec: PropertySpec) = apply {
      propertySpecs += propertySpec
    }

    fun addProperty(name: String, type: TypeName, optional: Boolean = false, vararg modifiers: Modifier)
       = addProperty(PropertySpec.builder(name, type, optional, *modifiers).build())

    fun addFunctions(functionSpecs: Iterable<FunctionSpec>) = apply {
      functionSpecs.forEach { addFunction(it) }
    }

    fun addFunction(functionSpec: FunctionSpec) = apply {
      require(!functionSpec.isConstructor) { "Use the 'constructor' method for the constructor" }
      this.functionSpecs += functionSpec
    }

    fun build(): ClassSpec {
      val isAbstract = modifiers.contains(Modifier.ABSTRACT)
      for (functionSpec in functionSpecs) {
        require(isAbstract || !functionSpec.modifiers.contains(Modifier.ABSTRACT)) {
          "non-abstract type $name cannot declare abstract function ${functionSpec.name}"
        }
      }

      return ClassSpec(this)
    }
  }

  companion object {

    @JvmStatic
    fun builder(name: String) = Builder(name)

    @JvmStatic
    fun builder(name: TypeName) = Builder(
       name.reference(null))

  }

}

File: src/ClassSpec.ts
import { imm, Imm } from 'ts-imm';
import { CodeBlock } from './CodeBlock';
import { CodeWriter } from './CodeWriter';
import { DecoratorSpec } from './DecoratorSpec';
import { Encloser, FunctionSpec } from './FunctionSpec';
import { Modifier } from './Modifier';
import { ParameterSpec } from './ParameterSpec';
import { PropertySpec } from './PropertySpec';
import { TypeName, TypeNameOrString, TypeNames, TypeVariable } from './TypeNames';

/** A generated `class` declaration. */
export class ClassSpec extends Imm<ClassSpec> {
  public static create(name: string | TypeName): ClassSpec {
    return new ClassSpec({
      name: typeof name === 'string' ? name : name.reference(),
      javaDoc: CodeBlock.empty(),
      decorators: [],
      modifiers: [],
      typeVariables: [],
      superClassField: undefined,
      interfaces: [],
      propertySpecs: [],
      constructorField: undefined,
      functionSpecs: [],
    });
  }

  @imm
  public readonly name!: string;
  @imm
  public readonly javaDoc!: CodeBlock;
  @imm
  public readonly decorators!: DecoratorSpec[];
  @imm
  public readonly modifiers!: Modifier[];
  @imm
  public readonly typeVariables!: TypeVariable[];
  @imm
  public readonly superClassField?: TypeName;
  @imm
  public readonly interfaces!: TypeName[];
  @imm
  public readonly propertySpecs!: PropertySpec[];
  @imm
  public readonly constructorField?: FunctionSpec;
  @imm
  public readonly functionSpecs!: FunctionSpec[];

  public emit(codeWriter: CodeWriter): void {
    const constructorProperties: Map<string, PropertySpec> = this.constructorProperties();

    codeWriter.emitJavaDoc(this.javaDoc);
    codeWriter.emitDecorators(this.decorators, false);
    codeWriter.emitModifiers(this.modifiers, [Modifier.PUBLIC]);
    codeWriter.emit('class');
    codeWriter.emitCode(' %L', this.name);
    codeWriter.emitTypeVariables(this.typeVariables);

    const sc = this.superClassField ? CodeBlock.of('extends %T', this.superClassField) : CodeBlock.empty();
    const interfaces = CodeBlock.joinToCode(this.interfaces.map(it => CodeBlock.of('%T', it)), ', ', 'implements ');
    if (sc.isNotEmpty() && interfaces.isNotEmpty()) {
      codeWriter.emitCode(' %L %L', sc, interfaces);
    } else if (sc.isNotEmpty() || interfaces.isNotEmpty()) {
      codeWriter.emitCode(' %L%L', sc, interfaces);
    }

    codeWriter.emit(' {\n');
    codeWriter.indent();

    // Non-static properties.
    this.propertySpecs.forEach(propertySpec => {
      if (!constructorProperties.has(propertySpec.name)) {
        codeWriter.emit('\n');
        propertySpec.emit(codeWriter, [Modifier.PUBLIC], true);
      }
    });

    // Write the constructor manually, allowing the replacement
    // of property specs with constructor parameters
    if (this.constructorField) {
      codeWriter.emit('\n');
      const it = this.constructorField;
      if (it.decorators.length > 0) {
        codeWriter.emit(' ');
        codeWriter.emitDecorators(it.decorators, true);
        codeWriter.emit('\n');
      }
      if (it.modifiers.length > 0) {
        codeWriter.emitModifiers(it.modifiers);
      }
      codeWriter.emit('constructor');

      let body = it.body;
      // Emit constructor parameters & property specs that can be replaced with parameters
      ParameterSpec.emitAll(it.parameters, codeWriter, true, it.restParameter, (param, isRest) => {
        let property = constructorProperties.get(param.name);
        if (property && !isRest) {
          // Ensure the parameter always has a modifier (that makes it a property in TS)
          if (
            !property.modifiers.find(m =>
              [Modifier.PUBLIC, Modifier.PRIVATE, Modifier.PROTECTED, Modifier.READONLY].includes(m)
            )
          ) {
            // Add default public modifier
            property = property.addModifiers(Modifier.PUBLIC);
          }
          property.emit(codeWriter, [], false);
          param.emitDefaultValue(codeWriter);

          // Remove initializing statements
          body = body.remove(this.constructorPropertyInitSearch(property.name));
        } else {
          param.emit(codeWriter, true, isRest);
        }
      });

      codeWriter.emit(' {\n');
      codeWriter.indent();
      codeWriter.emitCodeBlock(body);
      codeWriter.unindent();
      codeWriter.emit('}\n');
    }

    // Constructors.
    this.functionSpecs.forEach(funSpec => {
      if (funSpec.isConstructor()) {
        codeWriter.emit('\n');
        funSpec.emit(codeWriter, [Modifier.PUBLIC]);
      }
    });

    // Functions (static and non-static).
    this.functionSpecs.forEach(funSpec => {
      if (!funSpec.isConstructor()) {
        codeWriter.emit('\n');
        funSpec.emit(codeWriter, [Modifier.PUBLIC]);
      }
    });

    codeWriter.unindent();

    if (!this.hasNoBody) {
      codeWriter.emit('\n');
    }
    codeWriter.emit('}\n');
  }

  public addJavadoc(format: string, ...args: any[]): this {
    return this.copy({
      javaDoc: this.javaDoc.add(format, ...args),
    });
  }

  public addJavadocBlock(block: CodeBlock): this {
    return this.copy({
      javaDoc: this.javaDoc.addCode(block),
    });
  }

  public addDecorators(...decoratorSpecs: DecoratorSpec[]): this {
    return this.copy({
      decorators: [...this.decorators, ...decoratorSpecs],
    });
  }

  public addDecorator(decoratorSpec: DecoratorSpec): this {
    return this.copy({
      decorators: [...this.decorators, decoratorSpec],
    });
  }

  public addModifiers(...modifiers: Modifier[]): this {
    return this.copy({
      modifiers: [...this.modifiers, ...modifiers],
    });
  }

  public addTypeVariables(...typeVariables: TypeVariable[]): this {
    return this.copy({
      typeVariables: [...this.typeVariables, ...typeVariables],
    });
  }

  public addTypeVariable(typeVariable: TypeVariable): this {
    return this.copy({
      typeVariables: [...this.typeVariables, typeVariable],
    });
  }

  public superClass(superClass: TypeNameOrString): this {
    // check(this.superClass == null) { "superclass already set to ${this.superClass}" }
    return this.copy({
      superClassField: TypeNames.anyTypeMaybeString(superClass),
    });
  }

  public addInterfaces(interfaces: TypeName[]): this {
    return this.copy({
      interfaces: [...this.interfaces, ...interfaces],
    });
  }

  public addInterface(iface: TypeNameOrString): this {
    return this.copy({
      interfaces: [...this.interfaces, TypeNames.anyTypeMaybeString(iface)],
    });
  }

  // "constructor" can't be a method name
  public cstr(constructor?: FunctionSpec): this {
    if (constructor) {
      // require(constructor.isConstructor) { "expected a constructor but was ${constructor.name}; use FunctionSpec.createConstructor when building"
    }
    return this.copy({
      constructorField: constructor,
    });
  }

  public addProperties(...propertySpecs: PropertySpec[]): this {
    return this.copy({
      propertySpecs: [...this.propertySpecs, ...propertySpecs],
    });
  }

  public addProperty(propertySpec: PropertySpec): this;
  public addProperty(name: string, type: TypeName | string, data: Partial<PropertySpec>): this;
  public addProperty(): this {
    let propertySpec: PropertySpec;
    if (arguments[0] instanceof PropertySpec) {
      propertySpec = arguments[0];
    } else {
      const name = arguments[0];
      const type = TypeNames.anyTypeMaybeString(arguments[1]);
      const data = arguments[2] || {};
      propertySpec = PropertySpec.create(name, type).copy(data);
    }
    return this.copy({
      propertySpecs: [...this.propertySpecs, propertySpec],
    });
  }

  public addFunctions(...functionSpecs: FunctionSpec[]): this {
    functionSpecs.forEach(it => this.addFunction(it));
    return this;
  }

  public addFunction(functionSpec: FunctionSpec): this {
    // require(!functionSpec.isConstructor) { "Use the 'constructor' method for the constructor" }
    return this.copy({
      functionSpecs: [...this.functionSpecs, functionSpec.setEnclosed(Encloser.CLASS)],
    });
  }

  public toString(): string {
    return CodeWriter.emitToString(this);
  }

  /** Returns the properties that can be declared inline as constructor parameters. */
  private constructorProperties(): Map<string, PropertySpec> {
    const cstr = this.constructorField;
    if (!cstr || !cstr.body) {
      return new Map();
    }
    const result: Map<string, PropertySpec> = new Map();
    this.propertySpecs.forEach(property => {
      const parameter = cstr.parameter(property.name);
      if (
        !parameter ||
        parameter.type !== property.type ||
        parameter.optional !== property.optional ||
        property.initializerField !== undefined
      ) {
        return;
      }
      const foundAssignment = cstr.body.formatParts.find(
        p => p.match(this.constructorPropertyInitSearch(property.name)) !== null
      );
      if (!foundAssignment) {
        return;
      }
      result.set(property.name, property);
    });
    return result;
  }

  private constructorPropertyInitSearch(n: string): RegExp {
    // the outfoxx code was a lot fancier than this, but for now do the simple thing
    // and only match on standalone lines that are exactly `this.foo = foo` (we assume
    // any indentation and
    const pattern = `^this\\.${n}\\s*=\\s*${n}$`;
    return new RegExp(pattern);
  }

  private get hasNoBody(): boolean {
    if (this.propertySpecs.length > 0) {
      const constructorProperties = this.constructorProperties();
      const nonCstrProperties = this.propertySpecs.filter(p => !constructorProperties.has(p.name));
      if (nonCstrProperties.length > 0) {
        return false;
      }
    }
    return this.constructorField === undefined && this.functionSpecs.length === 0;
  }
}
